{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "radical-anxiety",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "from prettytable import PrettyTable\n",
    "import plotly.graph_objects as go\n",
    "import numpy as np\n",
    "import random\n",
    "import csv\n",
    "import os\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "import math\n",
    "from scipy.stats import wasserstein_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "outer-supplier",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Local on Windows 10 box\n",
    "WD = os.path.join(\"E:\\\\\", \"BUSTEDS-MH\")\n",
    "\n",
    "tag = \"empirical_unmasked_selectome\"\n",
    "\n",
    "# Additional declares\n",
    "BUSTEDS_DIR = os.path.join(WD, \"analysis\", tag, \"BUSTEDS\")\n",
    "BUSTEDS_MH_DIR = os.path.join(WD, \"analysis\", tag, \"BUSTEDS-MH\")\n",
    "\n",
    "OUTPUT_CSV = os.path.join(WD, \"tables\", \"Table_\" + tag.upper() + \"_BUSTEDS_and_BUSTEDS-MH.csv\")\n",
    "\n",
    "ER_Threshold = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "sealed-flexibility",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json(filename):\n",
    "    #print(\"# Reading:\", filename)\n",
    "    if os.stat(filename).st_size == 0: \n",
    "        print(\"# -- Error -- file is empty:\", filename)\n",
    "        return []\n",
    "    #end if\n",
    "    with open(filename, \"r\") as fh:\n",
    "        json_data = json.load(fh)\n",
    "    fh.close()\n",
    "    return json_data\n",
    "#end method\n",
    "\n",
    "#define function to calculate cv\n",
    "#cv = lambda x: np.std(x, ddof=1) / np.mean(x) * 100 \n",
    "cv = lambda x: np.std(x) / np.mean(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "given-atlanta",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Number of BUSTEDS results: 13307\n",
      "# Number of BUSTEDS-MH results: 13299\n"
     ]
    }
   ],
   "source": [
    "BUSTEDS_DIR_FILES = [os.path.join(BUSTEDS_DIR, file.name) for file in os.scandir(BUSTEDS_DIR) if file.name.endswith(\".json\")]\n",
    "BUSTEDS_MH_DIR_FILES = [os.path.join(BUSTEDS_MH_DIR, file.name) for file in os.scandir(BUSTEDS_MH_DIR) if file.name.endswith(\".json\")]\n",
    "\n",
    "print(\"# Number of BUSTEDS results:\", len(BUSTEDS_DIR_FILES))\n",
    "print(\"# Number of BUSTEDS-MH results:\", len(BUSTEDS_MH_DIR_FILES))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc7a7e6",
   "metadata": {},
   "source": [
    "## Look over BUSTEDS-MH Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d65cee32",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████| 13299/13299 [02:15<00:00, 98.27it/s]\n"
     ]
    }
   ],
   "source": [
    "df_dict = {}\n",
    "\n",
    "for item in tqdm(BUSTEDS_MH_DIR_FILES):\n",
    "    basename = os.path.basename(item).replace(\".nex.BUSTEDS-MH.json\", \"\")\n",
    "    # Read \n",
    "    json_data_BUSTEDS_MH = read_json(item)\n",
    "    \n",
    "    if json_data_BUSTEDS_MH == []: continue\n",
    "    \n",
    "    #print(\"# Data loaded\")\n",
    "    df_dict[basename] = {\"Method\": \"BUSTEDS-MH\"}\n",
    "    df_dict[basename].update({\"Sequences\": json_data_BUSTEDS_MH[\"input\"][\"number of sequences\"]})\n",
    "    df_dict[basename].update({\"Codons\": json_data_BUSTEDS_MH[\"input\"][\"number of sites\"]})\n",
    "    df_dict[basename].update({\"LRT p-value\": json_data_BUSTEDS_MH[\"test results\"][\"p-value\"]})\n",
    "\n",
    "    # cAIC\n",
    "    df_dict[basename].update({\"cAIC\": json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"AIC-c\"]})\n",
    "    \n",
    "    # CV of omega\n",
    "    #A = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"0\"][\"omega\"] \n",
    "    #B = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"1\"][\"omega\"] \n",
    "    #C = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"2\"][\"omega\"] \n",
    "    #df_dict[basename].update({\"CV(omega)\": cv([A, B, C])})\n",
    "    \n",
    "    # CV of alpha\n",
    "    #D = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"0\"][\"rate\"] \n",
    "    #E = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"1\"][\"rate\"] \n",
    "    #F = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"2\"][\"rate\"] \n",
    "    #df_dict[basename].update({\"CV(alpha)\": cv([D, E, F])})\n",
    "    \n",
    "    #Omegas and proportions\n",
    "    data = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"]\n",
    "    w1 = round(data[\"0\"][\"omega\"], 4)\n",
    "    p1 = round(data[\"0\"][\"proportion\"], 4)\n",
    "    w2 = round(data[\"1\"][\"omega\"], 4)\n",
    "    p2 = round(data[\"1\"][\"proportion\"], 4)\n",
    "    w3 = round(data[\"2\"][\"omega\"], 4)\n",
    "    p3 = round(data[\"2\"][\"proportion\"], 4)\n",
    "    df_dict[basename].update({\"w1\": w1, \"p1\": p1})\n",
    "    df_dict[basename].update({\"w2\": w2, \"p2\": p2})\n",
    "    df_dict[basename].update({\"w3\": w3, \"p3\": p3})\n",
    "    \n",
    "    df_dict[basename].update({\"CV(omega)\": cv([w1, w2, w3])})\n",
    "    \n",
    "    # SRV rates and proportions\n",
    "    data = json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"]\n",
    "    s1 = round(data[\"0\"][\"rate\"], 4)\n",
    "    s_p1 = round(data[\"0\"][\"proportion\"], 4)\n",
    "    s2 = round(data[\"1\"][\"rate\"], 4)\n",
    "    s_p2 = round(data[\"1\"][\"proportion\"], 4)\n",
    "    s3 = round(data[\"2\"][\"rate\"], 4)\n",
    "    s_p3 = round(data[\"2\"][\"proportion\"], 4)\n",
    "    df_dict[basename].update({\"SRV1\": s1, \"SRV_p1\": s_p1})\n",
    "    df_dict[basename].update({\"SRV2\": s2, \"SRV_p2\": s_p2})\n",
    "    df_dict[basename].update({\"SRV3\": s3, \"SRV_p3\": s3})\n",
    "    df_dict[basename].update({\"CV(alpha)\": cv([s1, s2, s3])})\n",
    "    \n",
    "    # DH rate, TH rate, TH_SI rate\n",
    "    df_dict[basename].update({\"DH_Rate\": float(json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"]\n",
    "                              [\"rate at which 2 nucleotides are changed instantly within a single codon\"])})\n",
    "    df_dict[basename].update({\"TH_Rate\": float(json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"]\n",
    "                              [\"rate at which 3 nucleotides are changed instantly within a single codon\"])})\n",
    "    df_dict[basename].update({\"TH_Rate_SI\": float(json_data_BUSTEDS_MH[\"fits\"][\"Unconstrained model\"]\n",
    "                              [\"rate at which 3 nucleotides are changed instantly within a single codon between synonymous codon islands\"])})\n",
    "\n",
    "    # ER Sites, thresholded\n",
    "    ER_SITES = []\n",
    "    ER_df_dict = {}\n",
    "    if \"constrained\" in json_data_BUSTEDS_MH[\"Evidence Ratios\"].keys():\n",
    "        #print(\"# ER Constrained Sites:\", len(json_data_BUSTEDS_MH[\"Evidence Ratios\"][\"constrained\"][0]))\n",
    "        for site, val in enumerate(json_data_BUSTEDS_MH[\"Evidence Ratios\"][\"constrained\"][0]):\n",
    "            if val > ER_Threshold:\n",
    "                ER_SITES.append(str(site + 1))\n",
    "                ER_df_dict[site + 1] = {\"BUSTEDS-MH ER\": val}\n",
    "            #end if\n",
    "        #end for\n",
    "        # add assert that there are more than 0 sites here.\n",
    "        df_dict[basename].update({\"BUSTEDS-MH_num_ER_Sites\":  len(ER_df_dict.keys())})\n",
    "        x = ER_df_dict.keys()\n",
    "        x = [str(x) for x in x]\n",
    "        df_dict[basename].update({\"BUSTEDS-MH_ER_Sites\":  \"|\".join(x)})\n",
    "        #print(ER_df_dict.keys())\n",
    "    #end if \n",
    "# end for\n",
    "\n",
    "df_MH = pd.DataFrame.from_dict(df_dict, orient=\"index\")\n",
    "df_MH = df_MH.reset_index()\n",
    "df_MH.index += 1\n",
    "df_MH.rename(columns={'index': 'Gene'}, inplace = True)\n",
    "#df_MH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37277f20",
   "metadata": {},
   "source": [
    "## Look over BUSTEDS Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b543d65a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████| 13307/13307 [02:32<00:00, 87.46it/s]\n"
     ]
    }
   ],
   "source": [
    "df_dict = {}\n",
    "\n",
    "for item in tqdm(BUSTEDS_DIR_FILES):\n",
    "    basename = os.path.basename(item).replace(\".nex.BUSTEDS.json\", \"\")\n",
    "    # Read json\n",
    "    #print()\n",
    "    json_data_BUSTEDS = read_json(item)\n",
    "    #print(\"# Data loaded:\", item)\n",
    "    \n",
    "    df_dict[basename] = {\"Method\": \"BUSTEDS\"}\n",
    "    df_dict[basename].update({\"Sequences\": json_data_BUSTEDS[\"input\"][\"number of sequences\"]})\n",
    "    df_dict[basename].update({\"Codons\": json_data_BUSTEDS[\"input\"][\"number of sites\"]})\n",
    "    df_dict[basename].update({\"LRT p-value\": json_data_BUSTEDS[\"test results\"][\"p-value\"]})\n",
    "\n",
    "    # cAIC\n",
    "    df_dict[basename].update({\"cAIC\": json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"AIC-c\"]})\n",
    "    \n",
    "    A = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"0\"][\"omega\"] \n",
    "    B = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"1\"][\"omega\"] \n",
    "    C = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"][\"2\"][\"omega\"] \n",
    "    df_dict[basename].update({\"CV(omega)\": cv([A, B, C])})\n",
    "    \n",
    "    D = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"0\"][\"rate\"] \n",
    "    E = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"1\"][\"rate\"] \n",
    "    F = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"][\"2\"][\"rate\"] \n",
    "    df_dict[basename].update({\"CV(alpha)\": cv([D, E, F])})\n",
    "    \n",
    "    #Omegas and proportions\n",
    "    data = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Test\"]\n",
    "    w1 = round(data[\"0\"][\"omega\"], 4)\n",
    "    p1 = round(data[\"0\"][\"proportion\"], 4)\n",
    "    w2 = round(data[\"1\"][\"omega\"], 4)\n",
    "    p2 = round(data[\"1\"][\"proportion\"], 4)\n",
    "    w3 = round(data[\"2\"][\"omega\"], 4)\n",
    "    p3 = round(data[\"2\"][\"proportion\"], 4)\n",
    "    df_dict[basename].update({\"w1\": w1, \"p1\": p1})\n",
    "    df_dict[basename].update({\"w2\": w2, \"p2\": p2})\n",
    "    df_dict[basename].update({\"w3\": w3, \"p3\": p3})\n",
    "    \n",
    "    # SRV rates and proportions\n",
    "    data = json_data_BUSTEDS[\"fits\"][\"Unconstrained model\"][\"Rate Distributions\"][\"Synonymous site-to-site rates\"]\n",
    "    s1 = round(data[\"0\"][\"rate\"], 4)\n",
    "    s_p1 = round(data[\"0\"][\"proportion\"], 4)\n",
    "    s2 = round(data[\"1\"][\"rate\"], 4)\n",
    "    s_p2 = round(data[\"1\"][\"proportion\"], 4)\n",
    "    s3 = round(data[\"2\"][\"rate\"], 4)\n",
    "    s_p3 = round(data[\"2\"][\"proportion\"], 4)\n",
    "    df_dict[basename].update({\"SRV1\": s1, \"SRV_p1\": s_p1})\n",
    "    df_dict[basename].update({\"SRV2\": s2, \"SRV_p2\": s_p2})\n",
    "    df_dict[basename].update({\"SRV3\": s3, \"SRV_p3\": s3})\n",
    "    \n",
    "    # ER Sites\n",
    "    ER_SITES = []\n",
    "    ER_df_dict = {}\n",
    "    \n",
    "    if \"constrained\" in json_data_BUSTEDS[\"Evidence Ratios\"].keys():\n",
    "        #print(\"# ER Constrained Sites:\", len(json_data_BUSTEDS[\"Evidence Ratios\"][\"constrained\"][0]))\n",
    "        for site, val in enumerate(json_data_BUSTEDS[\"Evidence Ratios\"][\"constrained\"][0]):\n",
    "            if val > ER_Threshold:\n",
    "                ER_SITES.append(str(site + 1))\n",
    "                ER_df_dict[site + 1] = {\"BUSTEDS ER\": val}\n",
    "            #end if\n",
    "        #end for\n",
    "        df_dict[basename].update({\"num_ER_Sites\":  int(len(ER_df_dict.keys()))})\n",
    "        df_dict[basename].update({\"BUSTEDS_num_ER_Sites\":  len(ER_df_dict.keys())})\n",
    "        x = ER_df_dict.keys()\n",
    "        x = [str(x) for x in x]\n",
    "        df_dict[basename].update({\"BUSTEDS_ER_Sites\":  \"|\".join(x)})\n",
    "        #print(ER_df_dict.keys())\n",
    "    #end if   \n",
    "    \n",
    "\n",
    "# End for\n",
    "\n",
    "df = pd.DataFrame.from_dict(df_dict, orient=\"index\")\n",
    "df = df.reset_index()\n",
    "df.index += 1\n",
    "df.rename(columns={'index': 'Gene'}, inplace = True)\n",
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "round-church",
   "metadata": {},
   "source": [
    "## Calculate cAIC statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ddc27f9d-a0b7-4798-8879-b142a16e3e04",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10781it [00:45, 220.15it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [Gene, Method, Sequences, Codons, LRT p-value, cAIC, CV(omega), CV(alpha), w1, p1, w2, p2, w3, p3, SRV1, SRV_p1, SRV2, SRV_p2, SRV3, SRV_p3, num_ER_Sites, BUSTEDS_num_ER_Sites, BUSTEDS_ER_Sites, ΔcAIC, RelativeSupport, ER_Sites_Intersection]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 26 columns]\n",
      "Series([], Name: w1, dtype: float64) Series([], Name: w2, dtype: float64) Series([], Name: w3, dtype: float64)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13299it [00:55, 238.97it/s]\n"
     ]
    }
   ],
   "source": [
    "df[\"ΔcAIC\"] = \"\"\n",
    "df[\"RelativeSupport\"] = \"\"\n",
    "df[\"ER_Sites_Intersection\"] = \"\"\n",
    "\n",
    "for index, row in tqdm(df_MH.iterrows()):\n",
    "    #MH_cAIC = df_MH[]\n",
    "    #df_temp = df[df[\"Gene\"] == gene]\n",
    "    #print(df_temp)\n",
    "    #print(row[\"Gene\"], row[\"cAIC\"])    \n",
    "    gene = row[\"Gene\"]\n",
    "    MH_cAIC = float(row[\"cAIC\"])\n",
    "    \n",
    "    BUSTEDS_row = df[df[\"Gene\"] == row[\"Gene\"]]\n",
    "    index_BUSTEDS = df[df[\"Gene\"] == row[\"Gene\"]].index\n",
    "    \n",
    "    try:\n",
    "        BUSTEDS_cAIC = float(BUSTEDS_row[\"cAIC\"])\n",
    "    except:\n",
    "        print(BUSTEDS_row)\n",
    "        pass\n",
    "    #end try\n",
    "    \n",
    "    BUSTEDSMH_w1, BUSTEDSMH_w2, BUSTEDSMH_w3 = float(row[\"w1\"]), float(row[\"w2\"]), float(row[\"w3\"])\n",
    "    try:\n",
    "        BUSTEDS_w1, BUSTEDS_w2, BUSTEDS_w3 = float(BUSTEDS_row[\"w1\"]), float(BUSTEDS_row[\"w2\"]), float(BUSTEDS_row[\"w3\"])\n",
    "    except:\n",
    "        print(BUSTEDS_row[\"w1\"], BUSTEDS_row[\"w2\"], BUSTEDS_row[\"w3\"])\n",
    "        continue\n",
    "    #end try\n",
    "    \n",
    "    BUSTEDSMH_p1, BUSTEDSMH_p2, BUSTEDSMH_p3 = float(row[\"p1\"]), float(row[\"p2\"]), float(row[\"p3\"])\n",
    "    BUSTEDS_p1, BUSTEDS_p2, BUSTEDS_p3 = float(BUSTEDS_row[\"p1\"]), float(BUSTEDS_row[\"p2\"]), float(BUSTEDS_row[\"p3\"])\n",
    "    \n",
    "    BUSTEDSMH_s1, BUSTEDSMH_s2, BUSTEDSMH_s3 = float(row[\"SRV1\"]), float(row[\"SRV2\"]), float(row[\"SRV3\"])\n",
    "    BUSTEDS_s1, BUSTEDS_s2, BUSTEDS_s3 = float(BUSTEDS_row[\"SRV1\"]), float(BUSTEDS_row[\"SRV2\"]), float(BUSTEDS_row[\"SRV3\"])\n",
    "    \n",
    "    BUSTEDSMH_sp1, BUSTEDSMH_sp2, BUSTEDSMH_sp3 = float(row[\"SRV_p1\"]), float(row[\"SRV_p2\"]), float(row[\"SRV_p3\"])\n",
    "    BUSTEDS_sp1, BUSTEDS_sp2, BUSTEDS_sp3 = float(BUSTEDS_row[\"SRV_p1\"]), float(BUSTEDS_row[\"SRV_p2\"]), float(BUSTEDS_row[\"SRV_p3\"])\n",
    "    \n",
    "    #print(float(BUSTEDS_cAIC[\"cAIC\"]))\n",
    "    best_model = min(MH_cAIC, BUSTEDS_cAIC)\n",
    "    #print()\n",
    "    #print(\"# Gene:\", row[\"Gene\"])\n",
    "\n",
    "    if BUSTEDS_cAIC == best_model:\n",
    "        which_is_best = \"BUSTEDS\"\n",
    "        delta_cAIC = MH_cAIC - best_model\n",
    "        relative_support = math.exp(-delta_cAIC/2)\n",
    "        # add to table\n",
    "        #df.at['C', 'x'] = 10\n",
    "        # \n",
    "        df.at[index_BUSTEDS, \"ΔcAIC\"] = delta_cAIC\n",
    "        df.at[index_BUSTEDS, \"RelativeSupport\"] = relative_support\n",
    "    elif MH_cAIC == best_model:\n",
    "        which_is_best = \"BUSTEDS-MH\"\n",
    "        delta_cAIC = BUSTEDS_cAIC - best_model\n",
    "        relative_support = math.exp(-delta_cAIC/2)\n",
    "        df_MH.at[index, \"ΔcAIC\"] = delta_cAIC\n",
    "        df_MH.at[index, \"RelativeSupport\"] = relative_support\n",
    "    else:\n",
    "        pass\n",
    "    #end if\n",
    "    #print(\"# Best model is:\", best_model, which_is_best, \"by\", delta_cAIC)\n",
    "    #print(\"# With relative support:\", relative_support)\n",
    "    \n",
    "    # Intersections of ER Sites.\n",
    "    #print(\"# Examining ER Sites\")\n",
    "    # BUSTEDS-MH_ER_Sites\n",
    "    # BUSTEDS_ER_Sites\n",
    "    try:\n",
    "        BUSTEDS_MH_ER_Sites = row[\"BUSTEDS-MH_ER_Sites\"].split(\"|\")\n",
    "        BUSTEDS_df = df[df[\"Gene\"] == row[\"Gene\"]]\n",
    "        BUSTEDS_ER_Sites    = BUSTEDS_df[\"BUSTEDS_ER_Sites\"].tolist()[0].split(\"|\")\n",
    "        #print(BUSTEDS_MH_ER_Sites, BUSTEDS_ER_Sites)\n",
    "        intersection = set(BUSTEDS_MH_ER_Sites).intersection(BUSTEDS_ER_Sites)\n",
    "        #print(intersection)\n",
    "        df.at[index, \"ER_Sites_Intersection\"] = \"|\".join(intersection)\n",
    "    except:\n",
    "        #print(\"ERROR --\", row[\"BUSTEDS-MH_ER_Sites\"])\n",
    "        pass\n",
    "    #end try\n",
    "    \n",
    "    \n",
    "    # Wasserstein Distance (Earth Movers)\n",
    "    #print(\"Omegas:\", BUSTEDSMH_w1, BUSTEDSMH_w2, BUSTEDSMH_w3, BUSTEDS_w1, BUSTEDS_w2, BUSTEDS_w3)\n",
    "    WD_unweighted_omega = wasserstein_distance([BUSTEDSMH_w1, BUSTEDSMH_w2, BUSTEDSMH_w3],\n",
    "                                               [BUSTEDS_w1, BUSTEDS_w2, BUSTEDS_w3])\n",
    "    \n",
    "    WD_weighted_omega   = wasserstein_distance([BUSTEDSMH_w1, BUSTEDSMH_w2, BUSTEDSMH_w3], \n",
    "                                               [BUSTEDS_w1, BUSTEDS_w2, BUSTEDS_w3],\n",
    "                                               [BUSTEDSMH_p1, BUSTEDSMH_p2, BUSTEDSMH_p3],\n",
    "                                               [BUSTEDS_p1, BUSTEDS_p2, BUSTEDS_p3])\n",
    "    \n",
    "    #print(\"SRV:\", BUSTEDSMH_s1, BUSTEDSMH_s2, BUSTEDSMH_s3, BUSTEDS_s1, BUSTEDS_s2, BUSTEDS_s3)\n",
    "    \n",
    "    WD_unweighted_srv   = wasserstein_distance([BUSTEDSMH_s1, BUSTEDSMH_s2, BUSTEDSMH_s3], \n",
    "                                               [BUSTEDS_s1, BUSTEDS_s2, BUSTEDS_s3])\n",
    "    \n",
    "    WD_weighted_srv     = wasserstein_distance([BUSTEDSMH_s1, BUSTEDSMH_s2, BUSTEDSMH_s3], \n",
    "                                               [BUSTEDS_s1, BUSTEDS_s2, BUSTEDS_s3], \n",
    "                                               [BUSTEDSMH_sp1, BUSTEDSMH_sp2, BUSTEDSMH_sp3], \n",
    "                                               [BUSTEDS_sp1, BUSTEDS_sp2, BUSTEDS_sp3])\n",
    "    \n",
    "    df_MH.at[index, \"WD_unweighted(omega)\"] = WD_unweighted_omega\n",
    "    df_MH.at[index, \"WD_weighted(omega)\"] = WD_weighted_omega\n",
    "    df_MH.at[index, \"WD_unweighted(srv)\"] = WD_unweighted_srv\n",
    "    df_MH.at[index, \"WD_weighted(srv)\"] = WD_weighted_srv\n",
    "    \n",
    "    \n",
    "#end for\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5509d43",
   "metadata": {},
   "source": [
    "## Concat tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "908b0296",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Method</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Codons</th>\n",
       "      <th>LRT p-value</th>\n",
       "      <th>cAIC</th>\n",
       "      <th>w1</th>\n",
       "      <th>p1</th>\n",
       "      <th>w2</th>\n",
       "      <th>p2</th>\n",
       "      <th>...</th>\n",
       "      <th>ΔcAIC</th>\n",
       "      <th>RelativeSupport</th>\n",
       "      <th>WD_unweighted(omega)</th>\n",
       "      <th>WD_weighted(omega)</th>\n",
       "      <th>WD_unweighted(srv)</th>\n",
       "      <th>WD_weighted(srv)</th>\n",
       "      <th>num_ER_Sites</th>\n",
       "      <th>BUSTEDS_num_ER_Sites</th>\n",
       "      <th>BUSTEDS_ER_Sites</th>\n",
       "      <th>ER_Sites_Intersection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSGT00390000000002.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>24</td>\n",
       "      <td>428</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>15230.383413</td>\n",
       "      <td>0.0083</td>\n",
       "      <td>0.8481</td>\n",
       "      <td>0.2237</td>\n",
       "      <td>0.1472</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSGT00390000000002.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS-MH</td>\n",
       "      <td>24</td>\n",
       "      <td>428</td>\n",
       "      <td>0.453178</td>\n",
       "      <td>15218.395700</td>\n",
       "      <td>0.0108</td>\n",
       "      <td>0.8748</td>\n",
       "      <td>0.2630</td>\n",
       "      <td>0.1247</td>\n",
       "      <td>...</td>\n",
       "      <td>11.987712</td>\n",
       "      <td>0.002494</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.018824</td>\n",
       "      <td>0.002633</td>\n",
       "      <td>0.004728</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSGT00390000000002.Euteleostomi.002</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>35</td>\n",
       "      <td>520</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>28163.498555</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>0.9297</td>\n",
       "      <td>0.2579</td>\n",
       "      <td>0.0703</td>\n",
       "      <td>...</td>\n",
       "      <td>4.460285</td>\n",
       "      <td>0.107513</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSGT00390000000002.Euteleostomi.002</td>\n",
       "      <td>BUSTEDS-MH</td>\n",
       "      <td>35</td>\n",
       "      <td>520</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>28167.958840</td>\n",
       "      <td>0.0288</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0487</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.114767</td>\n",
       "      <td>0.026983</td>\n",
       "      <td>0.012767</td>\n",
       "      <td>0.010951</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSGT00390000000002.Euteleostomi.003</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>30</td>\n",
       "      <td>381</td>\n",
       "      <td>0.342345</td>\n",
       "      <td>18555.324235</td>\n",
       "      <td>0.0113</td>\n",
       "      <td>0.4917</td>\n",
       "      <td>0.0967</td>\n",
       "      <td>0.4930</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26602</th>\n",
       "      <td>ENSGT00680000101373.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS-MH</td>\n",
       "      <td>6</td>\n",
       "      <td>438</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>3894.546024</td>\n",
       "      <td>0.0363</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3503</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.000967</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.1055</td>\n",
       "      <td>0.078059</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26603</th>\n",
       "      <td>ENSGT00680000101376.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>6</td>\n",
       "      <td>344</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>2828.089034</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.8940</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1060</td>\n",
       "      <td>...</td>\n",
       "      <td>6.190516</td>\n",
       "      <td>0.045263</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26604</th>\n",
       "      <td>ENSGT00680000101376.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS-MH</td>\n",
       "      <td>6</td>\n",
       "      <td>344</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>2834.279550</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.0038</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.001793</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26605</th>\n",
       "      <td>ENSGT00680000101377.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS</td>\n",
       "      <td>6</td>\n",
       "      <td>405</td>\n",
       "      <td>0.015805</td>\n",
       "      <td>4896.192723</td>\n",
       "      <td>0.1444</td>\n",
       "      <td>0.9866</td>\n",
       "      <td>0.1454</td>\n",
       "      <td>0.0032</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>371</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26606</th>\n",
       "      <td>ENSGT00680000101377.Euteleostomi.001</td>\n",
       "      <td>BUSTEDS-MH</td>\n",
       "      <td>6</td>\n",
       "      <td>405</td>\n",
       "      <td>0.241076</td>\n",
       "      <td>4891.992925</td>\n",
       "      <td>0.1493</td>\n",
       "      <td>0.9697</td>\n",
       "      <td>0.1501</td>\n",
       "      <td>0.0244</td>\n",
       "      <td>...</td>\n",
       "      <td>4.199799</td>\n",
       "      <td>0.122469</td>\n",
       "      <td>3.025467</td>\n",
       "      <td>0.152328</td>\n",
       "      <td>0.014167</td>\n",
       "      <td>0.018231</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26606 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Gene      Method  Sequences  Codons  \\\n",
       "1      ENSGT00390000000002.Euteleostomi.001     BUSTEDS         24     428   \n",
       "2      ENSGT00390000000002.Euteleostomi.001  BUSTEDS-MH         24     428   \n",
       "3      ENSGT00390000000002.Euteleostomi.002     BUSTEDS         35     520   \n",
       "4      ENSGT00390000000002.Euteleostomi.002  BUSTEDS-MH         35     520   \n",
       "5      ENSGT00390000000002.Euteleostomi.003     BUSTEDS         30     381   \n",
       "...                                     ...         ...        ...     ...   \n",
       "26602  ENSGT00680000101373.Euteleostomi.001  BUSTEDS-MH          6     438   \n",
       "26603  ENSGT00680000101376.Euteleostomi.001     BUSTEDS          6     344   \n",
       "26604  ENSGT00680000101376.Euteleostomi.001  BUSTEDS-MH          6     344   \n",
       "26605  ENSGT00680000101377.Euteleostomi.001     BUSTEDS          6     405   \n",
       "26606  ENSGT00680000101377.Euteleostomi.001  BUSTEDS-MH          6     405   \n",
       "\n",
       "       LRT p-value          cAIC      w1      p1      w2      p2  ...  \\\n",
       "1         0.500000  15230.383413  0.0083  0.8481  0.2237  0.1472  ...   \n",
       "2         0.453178  15218.395700  0.0108  0.8748  0.2630  0.1247  ...   \n",
       "3         0.500000  28163.498555  0.0171  0.9297  0.2579  0.0703  ...   \n",
       "4         0.500000  28167.958840  0.0288  1.0000  0.0487  0.0000  ...   \n",
       "5         0.342345  18555.324235  0.0113  0.4917  0.0967  0.4930  ...   \n",
       "...            ...           ...     ...     ...     ...     ...  ...   \n",
       "26602     0.500000   3894.546024  0.0363  0.0000  0.3503  1.0000  ...   \n",
       "26603     0.500000   2828.089034  0.0000  0.8940  0.0000  0.1060  ...   \n",
       "26604     0.500000   2834.279550  0.0000  1.0000  0.0006  0.0000  ...   \n",
       "26605     0.015805   4896.192723  0.1444  0.9866  0.1454  0.0032  ...   \n",
       "26606     0.241076   4891.992925  0.1493  0.9697  0.1501  0.0244  ...   \n",
       "\n",
       "           ΔcAIC  RelativeSupport  WD_unweighted(omega)  WD_weighted(omega)  \\\n",
       "1                                                                             \n",
       "2      11.987712         0.002494                  1.72            0.018824   \n",
       "3       4.460285         0.107513                                             \n",
       "4                                              0.114767            0.026983   \n",
       "5                                                                             \n",
       "...          ...              ...                   ...                 ...   \n",
       "26602                                          0.000967              0.0001   \n",
       "26603   6.190516         0.045263                                             \n",
       "26604                                            0.0038                 0.0   \n",
       "26605                                                                         \n",
       "26606   4.199799         0.122469              3.025467            0.152328   \n",
       "\n",
       "       WD_unweighted(srv)  WD_weighted(srv)  num_ER_Sites  \\\n",
       "1                                                     0.0   \n",
       "2                0.002633          0.004728                 \n",
       "3                                                           \n",
       "4                0.012767          0.010951                 \n",
       "5                                                     0.0   \n",
       "...                   ...               ...           ...   \n",
       "26602              0.1055          0.078059                 \n",
       "26603                                                       \n",
       "26604            0.000567          0.001793                 \n",
       "26605                                                 1.0   \n",
       "26606            0.014167          0.018231                 \n",
       "\n",
       "       BUSTEDS_num_ER_Sites  BUSTEDS_ER_Sites  ER_Sites_Intersection  \n",
       "1                       0.0                                           \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "5                       0.0                                           \n",
       "...                     ...               ...                    ...  \n",
       "26602                                                                 \n",
       "26603                                                                 \n",
       "26604                                                                 \n",
       "26605                   1.0               371                         \n",
       "26606                                                                 \n",
       "\n",
       "[26606 rows x 35 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df = df.sort_values(by=\"Sequences\", ascending=False)\n",
    "#df_MH = df_MH.sort_values(by=\"Sequences\", ascending=False)\n",
    "\n",
    "result = pd.concat([df_MH, df])\n",
    "result = result.fillna(\"\")\n",
    "result = result.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "result = result.reset_index(drop=True)\n",
    "result.index += 1\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c3737b31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'dfv = result\\ndfv = dfv[[\\'Gene\\', \\'Sequences\\', \\'Method\\', \\'Codons\\', \\'LRT p-value\\', \\'cAIC\\', \\'delta cAIC (best model)\\', \\'Relative support\\',\\n       \\'CV(omega)\\', \\'CV(alpha)\\', \\'omega_3\\', \\'proportion_3\\', \\'DH_Rate\\',\\n       \\'TH_Rate\\', \\'TH_Rate_SI\\', \\'num_ER_Sites\\']]\\n\\ndfv = dfv.fillna(\"\")\\ndfv = dfv.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\\ndfv = dfv.reset_index(drop=True)\\ndfv.index += 1\\n'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"dfv = result\n",
    "dfv = dfv[['Gene', 'Sequences', 'Method', 'Codons', 'LRT p-value', 'cAIC', 'delta cAIC (best model)', 'Relative support',\n",
    "       'CV(omega)', 'CV(alpha)', 'omega_3', 'proportion_3', 'DH_Rate',\n",
    "       'TH_Rate', 'TH_Rate_SI', 'num_ER_Sites']]\n",
    "\n",
    "dfv = dfv.fillna(\"\")\n",
    "dfv = dfv.sort_values(by=[\"Gene\", \"Method\"], ascending=True)\n",
    "dfv = dfv.reset_index(drop=True)\n",
    "dfv.index += 1\n",
    "\"\"\"\n",
    "\n",
    "#styled_table = result.style.background_gradient()\n",
    "#styled_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "appreciated-india",
   "metadata": {},
   "source": [
    "## Save table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "satellite-commerce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving results to: E:\\BUSTEDS-MH\\tables\\Table_EMPIRICAL_UNMASKED_SELECTOME_BUSTEDS_and_BUSTEDS-MH.csv\n"
     ]
    }
   ],
   "source": [
    "print(\"Saving results to:\", OUTPUT_CSV)\n",
    "result.to_csv(OUTPUT_CSV, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4583c5bb",
   "metadata": {},
   "source": [
    "## End of file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "packed-pilot",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note Negative delta LL are convergence problems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "raised-nickname",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lower AIC values indicate a better-fit model, and a model with a delta-AIC (the difference between the two AIC values being compared) of more than -2 is considered significantly better than the model it is being compared to"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "specialized-anger",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Earth Mover's (Kantorovich) distance between two distrbuitions if you want a single number"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
